crates/bevy_transform/src/systems.rs

```diff
@@ -89,100 +145,423 @@ pub fn propagate_transforms(
+                // over the root entity query. Queries iterate over disjoint entities, preventing
+                // mutable aliasing, and making this call safe.
+                #[expect(unsafe_code, reason = "Mutating disjoint entities in parallel")]
+                unsafe {
+                    propagate_descendants_unchecked(
+                        parent,
+                        parent_transform,
+                        children,
+                        &nodes,
+                        outbox,
+                        &queue,
+                        // Need to revisit this single-max-depth by profiling more representative
+                        // scenes. It's possible that it is actually beneficial to go deep into the
+                        // hierarchy to build up a good task queue before starting the workers.
+                        // However, we avoid this for now to prevent cases where only a single
+                        // thread is going deep into the hierarchy while the others sit idle, which
+                        // is the problem that the tasks sharing workers already solve.
+                        1,
+                    );
+                }
+            },
+        );
+        // Send all tasks in thread local outboxes *after* roots are processed to reduce the total
+        // number of channel sends by avoiding sending partial batches.
+        queue.send_batches();
+
+        // Spawn workers on the task pool to recursively propagate the hierarchy in parallel.
+        let task_pool = ComputeTaskPool::get_or_init(TaskPool::default);
+        task_pool.scope(|s| {
+            (0..task_pool.thread_num())
+                .for_each(|_| s.spawn(async { propagation_worker(&queue, &nodes) }));
+        });
+    }
+
+    /// A parallel worker that will consume processed parent entities from the queue, and push
+    /// children to the queue once it has propagated their [`GlobalTransform`].
+    #[inline]
+    fn propagation_worker(queue: &WorkQueue, nodes: &NodeQuery) {
+        #[cfg(feature = "std")]
+        let _span = bevy_log::info_span!("transform propagation worker").entered();
+
+        let mut outbox = queue.local_queue.borrow_local_mut();
+        loop {
+            // Try to acquire a lock on the work queue in a tight loop. Profiling shows this is much
+            // more efficient than relying on `.lock()`, which causes gaps to form between tasks.
+            let Ok(rx) = queue.receiver.try_lock() else {
+                core::hint::spin_loop(); // No apparent impact on profiles, but best practice.
+                continue;
+            };
+            // If the queue is empty and no other threads are busy processing work, we can conclude
+            // there is no more work to do, and end the task by exiting the loop.
+            let Some(mut tasks) = rx.try_iter().next() else {
+                if queue.busy_threads.load(Ordering::Relaxed) == 0 {
```

Comment:
See next comment: does this need to be `Ordering::Release`? I don't think so, but not sure.

Reply:
To be honest, I'm not really sure. I found the docs pretty unhelpful.

Reply:
> Possible values are [SeqCst], [Acquire] and [Relaxed].

Reply:
I read the docs for cpp, and I've adjusted this to be a bit more correct, I think. No measurable impact to perf.